pool/src/pool.rs

```diff
@@ -409,6 +409,81 @@ where
 		}
 	}
 
+	/// Attempt to deaggregate a transaction and add it to the mempool
+	pub fn deaggregate_and_add_to_memory_pool(
+		&mut self,
+		tx_source: TxSource,
+		tx: transaction::Transaction,
+		stem: bool,
+	) -> Result<(), PoolError> {
+		match self.deaggregate_transaction(tx.clone()) {
+			Ok(deaggragated_tx) => self.add_to_memory_pool(tx_source, deaggragated_tx, stem),
+			Err(e) => {
+				debug!(
+					LOGGER,
+					"Could not deaggregate multi-kernel transaction: {:?}", e
+				);
+				self.add_to_memory_pool(tx_source, tx, stem)
+			}
+		}
+	}
+
+	/// Attempt to deaggregate multi-kernel transaction as much as possible based on the content
+	/// of the mempool
+	pub fn deaggregate_transaction(
+		&self,
+		tx: transaction::Transaction,
+	) -> Result<Transaction, PoolError> {
+		// find candidates tx and attempt to deaggregate
+		match self.find_candidates(tx.clone()) {
+			Some(candidates_txs) => match transaction::deaggregate(tx, candidates_txs) {
+				Ok(deaggregated_tx) => Ok(deaggregated_tx),
+				Err(e) => {
+					debug!(LOGGER, "Could not deaggregate transaction: {}", e);
+					Err(PoolError::FailedDeaggregation)
+				}
+			},
+			None => {
+				debug!(
+					LOGGER,
+					"Could not deaggregate transaction: no candidate transaction found"
+				);
+				Err(PoolError::FailedDeaggregation)
+			}
+		}
+	}
+
+	/// Find candidate transactions for a multi-kernel transaction
+	fn find_candidates(&self, tx: transaction::Transaction) -> Option<Vec<Transaction>> {
+		// While the inputs outputs can be cut-through the kernel will stay intact
+		// In order to deaggregate tx we look for tx with the same kernel
+		let mut found_txs: Vec<Transaction> = vec![];
+
+		// Gather all the kernels of the multi-kernel transaction in one set
+		let kernels_set: HashSet<TxKernel> = tx.kernels.iter().cloned().collect::<HashSet<_>>();
+
+		// Check each transaction in the pool
+		for (_, tx) in &self.transactions {
+			let candidates_kernels_set: HashSet<TxKernel> =
+				tx.kernels.iter().cloned().collect::<HashSet<_>>();
+
+			let kernels_set_intersection: HashSet<&TxKernel> =
+				kernels_set.intersection(&candidates_kernels_set).collect();
+
+			// Consider the transaction only if all the kernels match
+			if kernels_set_intersection.len() == tx.kernels.len() {
```

Comment:
Looks good with the set intersection - one more question though... is it sufficient to compare the `len()` here?

Can we use a stricter check here and look for one being the subset of the other?
https://doc.rust-lang.org/beta/std/collections/struct.HashSet.html#method.is_subset